app/lib/features/calendar_sync/calendar_sync.dart

```diff
@@ -0,0 +1,283 @@
+import 'package:acter_flutter_sdk/acter_flutter_sdk_ffi.dart';
+import 'package:logging/logging.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+final _log = Logger('a3::calendar_sync');
+
+final bool isSupportedPlatform = Platform.isAndroid || Platform.isIOS;
+typedef IdMapping = (String acterId, String localId);
+
+class CalendarSyncFailed extends Error {}
+
+const rejectionKey = 'rejected_calendar_sync';
+const calendarSyncKey = 'calendar_sync_id';
+const calendarSyncIdsKey = 'calendar_sync_ids';
+
+// internal state
+
+// ignore: unnecessary_late
+late DeviceCalendarPlugin deviceCalendar = DeviceCalendarPlugin();
+ProviderSubscription<AsyncValue<List<EventAndRsvp>>>? _subscription;
+
+Future<bool> _isEnabled() async {
+  try {
+    return (await rootNavKey.currentContext!
+        .read(asyncIsActiveProvider(LabsFeature.deviceCalendarSync).future));
+  } catch (e, s) {
+    _log.severe('Reading current context failed', e, s);
+    return false;
+  }
+}
+
+T? _logError<T>(Result<T> result, String msg, {bool doThrow = false}) {
+  if (result.hasErrors) {
+    for (final err in result.errors) {
+      _log.severe('$msg ${err.errorCode}: ${err.errorMessage}');
+    }
+    if (doThrow) {
+      throw CalendarSyncFailed();
+    }
+  }
+  if (doThrow && result.data == null) {
+    throw CalendarSyncFailed();
+  }
+  return result.data;
+}
+
+Future<void> initCalendarSync({bool ignoreRejection = false}) async {
+  if (!await _isEnabled()) {
+    _log.warning('Calendar Sync disabled');
+    return;
+  }
+  if (!isSupportedPlatform) {
+    _log.warning('Calendar Sync not available on this device');
+    return;
+  }
+  final SharedPreferences preferences = await sharedPrefs();
+
+  final hasPermission = await deviceCalendar.hasPermissions();
+
+  if (hasPermission.data == false) {
+    if (!ignoreRejection && (preferences.getBool(rejectionKey) ?? false)) {
+      _log.warning('user previously rejected calendar sync. quitting');
+      return;
+    }
+    final requesting = await deviceCalendar.requestPermissions();
+    if (requesting.data == false) {
+      await preferences.setBool(rejectionKey, true);
+      _log.warning('user rejected calendar sync. quitting');
+      return;
+    }
+  }
+  // FOR DEBUGGING CLEAR Acter CALENDARS VIA:
+  // await clearActerCalendars();
+
+  final calendarId = await _getOrCreateCalendar();
+  // clear if it existed before
+  _subscription?.close();
+  // start listening
+  _subscription =
+      ProviderScope.containerOf(rootNavKey.currentContext!, listen: true)
+          .listen(
+    eventsToSyncProvider,
+    (prev, next) async {
+      if (!next.hasValue) {
+        _log.info('ignoring state change without value');
+        return;
+      }
+      // FIXME: we probably want to debounce this ...
+      await _refreshCalendar(calendarId, next.valueOrNull ?? []);
+    },
+    fireImmediately: true,
+  );
+}
+
+Future<void> _refreshCalendar(
```

Comment:
Skipping review of `_refreshCalendar` as not getting much idea of the code.

Reply:
a shame, as this is the main part of the code ;)